### DISCUSSION TITLE
Issues upgrading MPDF 6.1 to MPDF 8.x

### DISCUSSION THREAD

**Discussion by evdmeer** [mpdf/mpdf#1986] [General]
As part of a larger scale upgrade from a codebase from PHP 7 to 8 we are updating our external libraries as well.

Currently we are still using MPDF 6.1 to generate a bunch of PDF files, which still works under PHP 7, but is throwing a bunch of warnings, deprecations and errors under PHP 8.3.
We need to update to a newer MPDF library, but I am running into some issues.

The code we have relies on ->getX() and ->getY() calls (combined with SetXY()) to directly write content to the PDF, but those functions don't exist anymore in the latest MPDF versions.
I have looked through the documentation, but I can't find any references to any replacement functions. They seem to come from the underlying FPDF library?

Were those functions silently deprecated?
Is there a replacement for them, or do we need to completely rewrite our PDF generating code from scratch?

**Comment by finwe** [comment]
There is no function `getX` or `getY` in mPDF 6.1 (or in 6.0 and even in 5.7) that I can find. Nor is it documented anywhere in the documentation. So there was nothing to deprecate.

You could use internal variables `$mpdf->x` and `$mpdf->y`, but as also undocumented, they can be marked as private in future versions.

**Comment by evdmeer** [comment]
> There is no function `getX` or `getY` in mPDF 6.1 (or in 6.0 and even in 5.7) that I can find. Nor is it documented anywhere in the documentation. So there was nothing to deprecate.
> 
> You could use internal variables `$mpdf->x` and `$mpdf->y`, but as also undocumented, they can be marked as private in future versions.

That was why I was so confused as well, I could not find any reference to those functions in MPDF, only in FPDF. Although MPDF is (was?) based on FPDF, I assumed it would just redirect those calls to FPDF.

I guess I will have to dive deeper into this and see what has been going on here.

**Reply by finwe** [reply]
mPDF is build on top of FPDF, there is no internal reference to forward to. My guess is you use somehow altered mPDF codebase, different from distribution. It is possible that these functions were added to FPDF long after it was used to initialize mPDF.

**Reply by evdmeer** [reply]
After some digging, it seems that the getY() and getX() functions seem to have existed in MDF 5.3, which I also found references to in our codebase. 

The function would just return ->y

`function GetY()
{
    //Get y position
    return $this->y;
}`

I suspect this code has been using an even older version (5.3) of MPDF, which was also present in our codebase.
For now I will see what happens if I refactor those to just reference ->x and ->y directly, but I agree that that is a dangerous option since access to those variables is not guaranteed in the future.

Perhaps a more substantial rewrite will be needed to get this code up to spec with the latest versions of MPDF ...
